torch/optim/zeroptimizer.py

```diff
@@ -0,0 +1,524 @@
+    """
+
+    if isinstance(value, torch.Tensor):
+        return value.to(device, non_blocking=non_blocking)
+
+    if isinstance(value, (list, tuple)):
+        values = [_recursive_copy_to_device(val, non_blocking=non_blocking, device=device) for val in value]
+        return values if isinstance(value, list) else tuple(values)
+
+    if isinstance(value, container_abcs.Mapping):
+        return {
+            key: _recursive_copy_to_device(val, non_blocking=non_blocking, device=device) for key, val in value.items()
+        }
+
+    return value
+
+
+def _get_global_rank(group: Any, rank: int) -> int:
+    if group is dist.group.WORLD:
+        return rank
+    else:
+        global_rank = dist.distributed_c10d._get_global_rank(group, rank)  # type: ignore
+    return global_rank
+
+
+class ZeROptimizer(Optimizer):
+    """Wraps an arbitrary :class:`optim.Optimizer <torch.optim.Optimizer>`
+    optimizer and shards its state as described by ZeRO_.
+    ::
+
+        opt = ShardedOptimizer(params, optim=torch.optim.Adam, lr=0.01)
+
+    .. _ZeRO: https://arxiv.org/abs/1910.02054
+
+    We use a greedy algorithm to pack a number of parameters
+    at each rank. Each parameter belongs to a single rank and
+    is not divided among rank.
+
+    After each rank completed their parameter update, they broadcast
+    the new version of the parameters to all other ranks to synchronize
+    the parameters for next round forward/backward computation.
+
+    Args:
+        params (list of tensors):
+            parameters to be optimized
+    Keyword Args:
+        optim (torch.nn.Optimizer):
+            optimizer to shard
+        group (group):
+            torch.distributed group (default: group.WORLD)
+        broadcast_buffer_size (int):
+            the size of the buffer used to batch the small parameter tensors (default 128k).
+    """
+
+    #: The optimizer used for a given shard
+    optim: Optimizer
+
+    in_super_constructor: bool
+
+    def __init__(
+        self,
+        params: _params_t,
+        optim: Type[Optimizer],
+        group: Optional[Any] = None,
+        broadcast_buffer_size: int = 2 ** 17,
+        **default: Any,
+    ):
+        # Hold all the model params in the root .param_groups
+        self.in_super_constructor = True
+        super().__init__(params, default)
+        self.in_super_constructor = False
+
+        # Partition information. lazy evaluation, computed if requested
+        self._per_device_params: OrderedDict[
+            torch.device, List[List[Parameter]]
+        ] = OrderedDict()  # device, rank, params
+        self._param_rank: Dict[torch.Tensor, int] = {}
+        self._partition_parameters: List[List[dict]] = []
+
+        # Build the wrapped optimizer, responsible for a shard of the params
+        self.group = group if group is not None else dist.group.WORLD
+        self.world_size = dist.get_world_size(self.group)
+        self.rank = dist.get_rank(self.group)
```

Comment:
there is a `self.grou.rank()` API:

https://github.com/pytorch/pytorch/blob/b5662ba0f0cb6d99805a44dd5b6bcda128468c83/torch/csrc/distributed/c10d/init.cpp#L595